image-definitions/check.sh

```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+#
+# Check if repositories were updates since patch and log it
+#
+# Usage: check.sh
+#
+
+# Kubeflow repository check
+echo "Check Kubeflow repository"
+REPO_DIR="kubeflow"
+cd $REPO_DIR
+KF_REPO_COMMIT=$(eval "git rev-parse --short HEAD")
+KF_PATCH_COMMIT=$(eval "cat ../kubeflow-patch-commit.txt")
+if [ $KF_REPO_COMMIT != KF_PATCH_COMMIT ]; then
+    echo "WARNING: Repository $REPO_DIR/ has been updated since patch was generated"
+    echo "Latest commit $KF_REPO_COMMIT Patch commit $KF_PATCH_COMMIT"
+fi
```

Comment:
I think we should return a non-zero exit code here instead of just echo.  This gets used in the middle of the tox images-setup env - it would be easy for a user to not spot this message in the middle of that environment execution and then build different images than they expected.

Reply:
Actually I wanted to be just a warning and not exit with error. What if I just want to rebuild regardless of when patch was created?
Example:
- Fixes to some Makefile and requirements.in to solve some CVE.
- Rebuild, scan, publish.
- Commit patch. (at this point commit ID = A)
Later on, may commits later that do not touch above Makefile and requirements.in:
- Bi-weekly scan will checkout at commit ID = B
- Apply patch (successfully, because Makefile and requirements.in were not touched, i.e. there was no conflicts)
- Rebuild, scan

Reply:
Yeah rebuilding regardless is an important use case, but I think both cases need to be handled appropriately.  If someone sees a `check.sh` script that is used in a sequence of `this_command && check.sh && that_command` they'll expect the check can fail the sequence.  As is, this is too easy for someone to pull the repo, run `tox -e images-setup` and think "it ran successfully, so I must have a repo set up that is the same as the git hash that's been committed".  That feels like a silent error that will surprise someone

Reply:
perhaps a command line argument that lets someone ignore the error when they explicitly say they want to? That way an intentional user has the flexibility you want, but nobody can be surprised by something they dont expect

Reply:
I can remove `check.sh` out of the workflow, it is not needed. Just wanted to have a warning log when building with tox. We can describe the possibility of difference in commits in README and that would be considered a warning to user.
In the workflow that is implemented in tox there will be no exit with error anyway. Adding options that are not used is not a good practice.
Let me redesign the workflow a bit to make less confusing.

Reply:
Separated sources setup and patch. Patching is a manual process in our case. Having in tox will help for readability, but caution should be taken. We cannot make this patching automatic. If the requirement is to make patching seamless, the better approach is not to have the whole image building workflow in tox.

Reply:
BTW, if patch fails that will cause an error and exit. Then warning messages will be available to indicate that change in commit happened. But if change in commit has happened and patch applied successfully, IMO a warning is enough.

Reply:
I like the idea of `check.sh` warning people, it just doesn't feel strong enough.  This feels like something I wouldn't expect to need to keep an eye out for if I didn't know about it ahead of time.

If I see a repo that stores:
* a git hash
* a patch file

and provides me with automation to pull a repo, apply a patch, and build an image, I expect that it is building the same image that was built previously using the same instructions.  Agreed that sometimes I might want to use the old procedure to build a more up to date image, but I don't think that's what someone will expect coming in without other context.

Put another way, this feels a bit like having a requirements.txt file that pins to `myPackage==1.0.0`, but an `install.sh` script that will just install the most recent `myPackage` rather than the pinned one.  The user might want to update to a new version of `myPackage`, but if there's a pinned version I think they'd expect to get that same pinned version

Maybe someone else can break the logjam.  @misohu what do you think?

Reply:
That's why I separated setup and patch in tox.
Setup brings up the latest.
Patch will try to apply the patch to the latest.
User has option to checkout pre-patch version using commit ID stored in `kubeflow-patch-commit.txt` (may be it should be called `kubeflow-pre-patch-coommit.txt`?) and anything they want with the source.
As I said, if this is confusing, we can remove `check.sh` from the equation and all the checks would need to be done by user manually. This will remove "unexpected" behaviour from workflow, because it will force use not to blindly run `tox image-patch`, but actually go to the source code and investigate. We cannot automate image definition workflow. Looks like we are trying to do it in `tox.ini`. I think, it is wrong place to do it.

Reply:
Ok took me sometime to understand what you guys fighting about. I like the idea of check.sh ... at first glance It did not even came to my mind that maybe the upstream code changed and I am actually building something maybe not relevant. So I like the check.sh I agree with @ca-scribner if it is a check it should end with error if check fails. We may have like two tox environments one for custom build without check and one with the check but I think that check should fail. 

One more thing in README there is not a single word about check anywhere so its quite confusing. Also I dont like the name of the file check. It is checking the latest commit in upstream and nothing else so we should rename it to something fancy (if we want to keep it). love you guys <3

Reply:
Here is the recap of scenarios that I hope will clarify my point: if check returns error, the whole workflow is useless and won't produce updated container images.

Goal produce container images based on changes in upstream and our patch. If patch does not apply, perform needed manual work and produce new patch.

1. Check logs warning.

a.
```
checkout upstream (no new commits) -> check commit (no warning log) -> apply patch -> patch succeeded -> build new updated image with patch
```
b.
```
checkout upstream (new commits) -> check commit (warning log) -> apply patch -> patch succeeded -> build new updated image with patch
```
c.
```
checkout upstream (new commits) -> check commit (warning log) -> apply patch -> patch failed -> do manual work and produce/apply new patch -> build new updated image with patch
```

2. Check returns error (same scenarios as above)
a.
```
checkout upstream (no new commits) -> check commit (no warning log) -> apply patch -> patch succeeded -> build new updated image with patch
```
b.
```
checkout upstream (new commits) -> check commit -> error and exit
manually investigate -> apply patch -> patch succeeded -> build new updated image with patch
```
c.
```
checkout upstream (new commits) -> check commit -> error and exit
manually investigate -> do manual work and produce/apply new patch -> build new updated image with patch
```

Reply:
Re-designed. @ca-scribner @misohu Please take a look. Two options for patching. One that errors and another one that forces the patch. The latter one is for powerusers and first one is for everyone else.

Reply:
I think part of the disagreement here is that we have slightly different goals in mind.  I recognise the primary goal is to automate the building of new images so we always have an as-up-to-date-as-possible image on the shelf, but imo a secondary (and almost free) goal would be to be able to recreate the images we previously released.  Without the second function, it feels a bit like we have source code in a repository but no way to actually compile it again.  

I think a simple way to achieve everything would be to add a command line arg to `check.sh`, something like `check.sh --warn-on-error` or similar.  Then when a human is trying to rebuild the past image, they have a `check.sh` that fails on a difference, whereas the CI that always wants to build for the future uses the arg to suppress the errors and build anyway.  That, or the CI that is publishing for the future can just not invoke `check.sh` at all.

Reply:
The latest change should take care of it. `tox -e image-patch` will fail if there is an update to upstream repo and won't apply the patch.